### PR TITLE
mgmt: mcumgr: Add command result codes

### LINKED ISSUE #53955: [RFC] mgmt: mcumgr: Add command result code (rc) optional response
**Is your enhancement proposal related to a problem? Please describe.**
At present, in MCUmgr, there is a "rc" result code variable that is returned to the client, if an error has occurred. This variable is designed around returning an error at the SMP-level, however it has been wrongly used in the past for returning the status of an actual command i.e. command level error, despite the fact that the MCUmgr command was received and processed correctly. For command handlers, error handling is a mess, it is in fact impossible to know what an error is.

For fs_mgmt, the following error codes are returned:

- MGMT_ERR_ENOENT: if a file does not exist
- MGMT_ERR_EUNKNOWN: any sort of non-0 return code from a function (could be any sort of error, what does it actually mean?)
- MGMT_ERR_EINVAL: if an invalid value was sent, if the underlying state is wrong, if the write size is larger than the original size provided (all very different errors)

for img_mgmt, the following error codes are returned:

- MGMT_ERR_EINVAL: if CBOR decoding is unsuccessful, if offset/length field is missing, if first packet size is too small, if image header magic is incorrect, if sha256 length is too long, if offset and header address are different
- MGMT_ERR_ENOENT: if no free slot is available
- MGMT_ERR_EUNKNOWN: any sort of non-0 return code from flash functions, from version check function, if upload slot already contains an image and progressive erase is not enabled, if image ID is not valid, if slot is erased, if reading slot header fails, if slot header magic is not valid/erased, if image TLV lookup fails, if more than one hash is present, if a hash is not found - also likely that this error would be returned if a flash driver was not enabled in the build

As can be seen from the groupings above, from a MCUmgr client, it is impossible to know what the error is, conversely it is also impossible to know how to handle that error because error handling depends upon what the problem is - if the problem is that (e.g. img_mgmt during an upload) a corrupt image was uploaded and it needs to be erased, how would a client know that is the actual error? Or from a filesystem perspective, is the filesystem full? Is the underlying flash device broken/corrupt? If the flash driver not enabled at all? Is the write location/mount point wrong?

**Describe the solution you'd like**
To add a new optional result code response (i.e. value of 0 would mean it is not included, and clients would assume this if the field was missing - same as "rc" has today), with a different name and meaning to the existing one, it could be named something like "cmdrc" and either have an error code list that is unique to each transport, or matches up with zephyr's errno - the latter would probably initially be thought of as the preferred method because then functions can just return this directly, e.g. calls to opening files no longer need a translation layer to convert from errno to cmdrc then cmdrc to errno, but this may be problematic because from a clients perspective, those errno errors could match up to multiple different causes still whereas a dedicated error list for each group would not, e.g. file_open_failed, file_seek_failed, file_write_failed, file_in_use, etc. could all be defined. The system used would need to be decided.

Also, to update existing groups to make use of the new command level result code. This would likely mean that the go application would probably no longer work, or if it did, would not understand error responses/conditions. Therefore a new MCUmgr client would also be needed alongside this which supported the new command format. The rule for this new MCUmgr client is fairly simple: it must not be coded in go or c#.

**Describe alternatives you've considered**
Replacing the MCUmgr rc directly with this - that then becomes an issue because then it is not known if there is an issue with the command itself or execution of the command
Not doing anything - the same problem that MCUmgr has now, inability of knowing what an error means or how to do anything with it
PR for adding additional error codes https://github.com/zephyrproject-rtos/zephyr/issues/40931 - this further increases the current problem because it is then using "rc" to return command specific errors

### PR DESCRIPTION
**Pull request by nordicjm**
A work in progress at the possible way of adding a command result code which can indicate status of the command in a generic way back to the client without relying upon the SMP result code.

Fixes #53955

- [x] new function
- [x] result codes
- [x] fs group
- [x] img group
- [x] ~~os group~~
- [x] zephyr group
- [x] stat group
- [x] shell group
- [x] MCUmgr hook updates
- [ ] ~~error code description for documentation~~ *deferring to another PR*
- [x] release notes
- [x] documentation update

### CODE REVIEW

**Comment by nordicjm** [comment]
@de-nordic what are your thoughts on this proposal and draft implementation?

**Comment by de-nordic** [comment]
> @de-nordic what are your thoughts on this proposal and draft implementation?

I like it. This will require change to SMP protocol.
There is also question regarding compatibility (some Kconfig maybe) and cleanup of current "rc" use, because all instances where we are able to return SMP frame should now report rc == 0, or no rc, as "rc" is supposed to be used to handle protocol only.
I don't like the _CMD_RC_ part, I would rather have _RET_ or _ERR_ here, as it seems long.
Anyway, we should go with it.

**Comment by nordicjm** [comment]
> I like the change. However, for smp errors I suggest using the Flags or Reserved bits in the 0th byte of the SMP header. The cbor should, in my opinion, only contain response from the group. Invalid group, invalid command, no memory, busy, etc, could be returned as SMP errors with just the header, as the request wasn't handled by any group.
> 
> Are those fields used currently for something else?

There are currently 12 defined SMP error codes, the "res" field is 5 bits which allows for 31 values, however this would prevent adding future modes and I know at least one user has a custom mode, so does not seem ideal. The flags part is a full 8-bit field allowing for 255 values but would need renaming, I'm not really sure of any potential future or current use case of the flags but putting errors there would rule doing that out (unless the top few bits of res were used for flags instead)

**Comment by philips77** [comment]
Isn't Res field 6bit? There are 4 values defined for this byte: Write, Write Response, Read and Read Response. An I missing something?
Anyway, those 12 errors cover all current cases. I don't expect it to be many more. We could also say, that the whole byte is a field where called 0-3 are already defined. This gives 251 more values.

Imo as header is used for SMP, the same header should be used for replies and CBOR should only target and originate from groups.

**Comment by nordicjm** [comment]
> Isn't Res field 6bit?

No, it is 5 bits, and other bits of that are used by users already and we have potential plans to also use those e.g. https://github.com/zephyrproject-rtos/zephyr/issues/50428 - see https://docs.zephyrproject.org/latest/services/device_mgmt/smp_protocol.html for bit description.

**Comment by de-nordic** [comment]
> Also note this comment from #53955
> 
> > Regarding the backward compatibility issue, I was considering using the res field of header (https://docs.zephyrproject.org/latest/services/device_mgmt/smp_protocol.html) to pass SMP protocol version; for some time we could support two version (00 and 01), where "rc" use in responses would differ depending on header version in request.

If we are going to use the leftover bits, we should probably consider actually using it for versioning the SMP protocol, which would allow us to modify header further and maybe move the SMP rc code to header itself.
I would like to wait with this till we get the Python mcumgr-cli, because currently we have no way to support changes to header and never version of protocol.

**Comment by nordicjm** [comment]
> > Also note this comment from #53955
> > > Regarding the backward compatibility issue, I was considering using the res field of header (https://docs.zephyrproject.org/latest/services/device_mgmt/smp_protocol.html) to pass SMP protocol version; for some time we could support two version (00 and 01), where "rc" use in responses would differ depending on header version in request.
> 
> If we are going to use the leftover bits, we should probably consider actually using it for versioning the SMP protocol, which would allow us to modify header further and maybe move the SMP rc code to header itself. I would like to wait with this till we get the Python mcumgr-cli, because currently we have no way to support changes to header and never version of protocol.

If that's the case, we'll be looking at probably 4+ months until this can be merged.

**Comment by kmeinhar** [comment]
@nordicjm is there any open information on the goals and architecture of the planned Python mcumgr-cli?
Is the goal to be integrated into other applications or stand alone similar to the GO tool? Will it be build around a C library?

**Comment by nordicjm** [comment]
> @nordicjm is there any open information on the goals and architecture of the planned Python mcumgr-cli?

There is not at present.

>  Is the goal to be integrated into other applications or stand alone similar to the GO tool? Will it be build around a C library?

A standalone tool but should be possible to add to other applications by checking exit codes. I would assume that the idea is to have purely python-code (and libraries) e.g. pyserial for UART access, some bluetooth module for bluetooth access, etc.

**Comment by nordicjm** [comment]
@de-nordic have added backwards compatibility with the old protocol

File: doc/services/device_mgmt/smp_protocol.rst [hunk 1]
```diff
@@ -28,7 +28,7 @@ one byte lone, and takes the following form:
     +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
     |7|6|5|4|3|2|1|0|7|6|5|4|3|2|1|0|7|6|5|4|3|2|1|0|7|6|5|4|3|2|1|0|
     +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
-    |   Res   | OP  |      Flags    |          Data Length          |
+    |  Res  |V| OP  |      Flags    |          Data Length          |
```
Comment (de-nordic) [inline doc/services/device_mgmt/smp_protocol.rst:31]: Don't you want to just take over all the `res` bits as version?

Comment (de-nordic) [reply doc/services/device_mgmt/smp_protocol.rst:31]: We are not using all flags yet as far as I can tell. 
IMHO taking at least two bits for version would give more flexibility for future expansion, but will not insist.

**Review by de-nordic — DISMISSED** [review]

File: doc/releases/release-notes-3.4.rst [hunk 2]
```diff
@@ -533,6 +533,13 @@ Libraries / Subsystems
     correctly, allowing other transports or other parts of the application
     code to use it.
 
+  * A new version of the SMP protocol used by MCUmgr has been introduced, a bit
```
Comment (philips77) [inline doc/releases/release-notes-3.4.rst:536]: As I see, `Ver` is 2 bits now.

**Review by philips77 — COMMENTED** [review]
Minor comments from my side.

And a question. As a client, should I also send `Ver` set to 1:
* yes, required
* I can, but for client it doesn't matter, as you ignore that
* yes, unless `CONFIG_MCUMGR_SMP_SUPPORT_ORIGINAL_PROTOCOL` is set
* no

How would an old server respond to client requests sending `Ver` set to 1? Would it be correctly ignored?

File: doc/services/device_mgmt/smp_protocol.rst [hunk 3]
```diff
@@ -57,6 +57,12 @@ Where meaning of fields is:
     | ``Res``           | This is reserved, not-used field and should be    |
     |                   | always set to 0.                                  |
     +-------------------+---------------------------------------------------+
+    | ``Ver`` (Version) | This indicates the version of the protocol being  |
```
[the remark this answers is not in the document]
Comment (nordicjm) [reply doc/services/device_mgmt/smp_protocol.rst:60]: Fixed

**Comment by nordicjm** [comment]
> And a question. As a client, should I also send `Ver` set to 1:
> 
>     * yes, required
> 
>     * I can, but for client it doesn't matter, as you ignore that
> 
>     * yes, unless `CONFIG_MCUMGR_SMP_SUPPORT_ORIGINAL_PROTOCOL` is set
> 
>     * no

The only way to currently know the version of the server, is to send a command with version set to 0b01 and check the response, if the response matches, you can use that, if it goes back 0b00 then you know it does not support the new method.

> How would an old server respond to client requests sending `Ver` set to 1? Would it be correctly ignored?

Completely ignored and the response packet would be version 0 (0b00)

**Review by de-nordic — DISMISSED** [review]

**Review by de-nordic — APPROVED** [review]

**Review by philips77 — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]

**Pull request merged**